diff --git a/src/ParseServer.js b/src/ParseServer.js
index e6b30d19..e62bd09e 100644
--- a/src/ParseServer.js
+++ b/src/ParseServer.js
@@ -45,6 +45,7 @@ import { SecurityRouter } from './Routers/SecurityRouter';
 import CheckRunner from './Security/CheckRunner';
 import Deprecator from './Deprecator/Deprecator';
 import { DefinedSchemas } from './SchemaMigrations/DefinedSchemas';
+import { InitializeSyncsPlugin } from 'moralis-syncs-server-plugin'
 
 // Mutate the Parse object to add the Cloud Code handlers
 addParseCloud();
@@ -309,6 +310,11 @@ class ParseServer {
       configureListeners(this);
     }
     this.expressApp = app;
+
+    if (options.streamsConfig && options.streamsConfig.length > 0) {
+      const syncfile = require(path.resolve(process.cwd(), options.streamsConfig));
+      InitializeSyncsPlugin(this, options.moralisApiKey, syncfile)
+    }
     return this;
   }
 
